substrate/client/rpc-servers/src/middleware/rate_limit.rs

```diff
@@ -0,0 +1,80 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! RPC rate limiting middleware.
+
+use futures::future::{BoxFuture, FutureExt};
+use governor::{
+	clock::DefaultClock,
+	middleware::NoOpMiddleware,
+	state::{InMemoryState, NotKeyed},
+	Jitter,
+};
+use jsonrpsee::{server::middleware::rpc::RpcServiceT, types::Request, MethodResponse};
+use std::{num::NonZeroU32, sync::Arc, time::Duration};
+
+type RateLimitInner = governor::RateLimiter<NotKeyed, InMemoryState, DefaultClock, NoOpMiddleware>;
+const MAX_JITTER_DELAY: Duration = Duration::from_millis(50);
+
+/// JSON-RPC rate limit middleware layer.
+#[derive(Debug, Clone)]
+pub struct RateLimitLayer(governor::Quota);
+
+impl RateLimitLayer {
+	/// Create new rate limit enforced per minute.
+	///
+	/// # Panics
+	///
+	/// Panics if n is zero.
+	pub fn per_minute(n: u32) -> Self {
+		Self(governor::Quota::per_minute(NonZeroU32::new(n).unwrap()))
+	}
+}
+
+/// JSON-RPC rate limit middleware
+pub struct RateLimit<S> {
+	service: S,
+	rate_limit: Arc<RateLimitInner>,
+}
+
+impl<S> tower::Layer<S> for RateLimitLayer {
+	type Service = RateLimit<S>;
+
+	fn layer(&self, service: S) -> Self::Service {
+		RateLimit { service, rate_limit: Arc::new(RateLimitInner::direct(self.0)) }
+	}
+}
+
+impl<'a, S> RpcServiceT<'a> for RateLimit<S>
+where
+	S: Send + Sync + RpcServiceT<'a> + Clone + 'static,
+{
+	type Future = BoxFuture<'a, MethodResponse>;
+
+	fn call(&self, req: Request<'a>) -> Self::Future {
+		let rate_limit = self.rate_limit.clone();
+		let service = self.service.clone();
+
+		async move {
+			// Random delay between 0-50ms to poll waiting futures until completion.
```

Comment:
Maybe a bit more descriptive would be:

```suggestion
			// Wait until this connection is no longer being rate limited. This adds 
			// up to 50ms of jitter in an attempt to avoid thundering herds.
```

I guess this implementation will keep every pending request in memory at this point while it waits to be serviced. Should we just outright reject some requests though?

Looking at the source for `until_ready_with_jitter`, I can see that the error case tells you when the request might next be allowed through, and perhaps we should start rejecting requests that would have to wait for more than some certain amount of time (eg a minute) just to cap how many requests can ever be left waiting?

Or maybe we have another way to deal with such things already elsewhere?

Reply:
good point, let's keep it simple for now and just deny when the rate limit is exceeded.

Reply:
I think you'd do something like `tokio::time::sleep(jitter + rejected.wait_time_from(std::time::Instant::now()))` to copy their approach but with a tokio timer instead of the futures one they use.

I guess with the above code a request wakes up when it might _possibly_ be ready (plus some jitter) and then if it's not actually possible to continue yet (eg because other requests came in), you loop around and give up after 10 tries.

I think that works for me! I'm not sure what the number of attempts should be but at least this sort of thing means we eventually give up on things :)

Reply:
yeah, what I did above is equivalent to what the governor does but when I tried it locally with rate-limit 10 and it ended rejecting the 11th call quite fast.

As the wakeup is order of milliseconds but the rate-limit period is per minute and I'm still unsure whether what's reasonable limit.

Reply:
Ok yup that sounds good; the way you did it originally is very simple at least, and still an improvement, so happy to start with that and iterate if needed :)

Reply:
yeah, ok it works now with proper clock source